### DISCUSSION TITLE
[@types/react] Preserve generic type parameters for component when passed to "lazy" or "memo"

### DISCUSSION THREAD

**Discussion by KurtGokhan** [DefinitelyTyped/DefinitelyTyped#69474] [Issues with a @types package]
`forwardRef` is going to be unnecessary in React 19 and it made the community happy. One of the reasons `forwardRef` is disliked is because it doesn't preserve generic type parameters of a component. Suppose a component like this:

```tsx
function MyCmp<T extends ReactNode>({ value }: { value: T }) {
  return <>{value}</>;
}
```

When this component is wrapped in a HOC like `forwardRef`, `memo` or `lazy`, the component loses the generic ability.

See this behavior in [TS playground](https://www.typescriptlang.org/play/?jsx=4&ts=5.5.0-beta#code/JYWwDg9gTgLgBAbziApiCAaOAlFBDAYxgDkIATFOAXzgDMoIQ4ByKfI5gKG9oFcA7IsAj84AWQCeAYXAAeACpwUADxgp+ZAM452JcigB8ACiQA3PABtelKgC5Ecc1ZT3FVAJSJOcOGxi8oUVkDBCdrKlkAegMAbk4qbgIRTXgxNAgUMhkwOABeZHSjSWz3OO5IyLgAdWgAa208bRUwFCJMziT+FKULAEY8uFliuX5eEAAjFCgDR0trXIQAVhposs4KuAARDM1+ZngAdzq4RqVlFrayDuT4FAsAJgGh9Mzs2VGJqZmwlAXmAAs7hYIMwVrFOEA).

### Proposed solution

The culprit is the `ExoticComponent` type. It converts any component to a simple component like `(props: P) => ReactNode` without preserving the generic type parameters.

There is a simple solution to this issue. See a POC of the proposed solution giving off the expected behavior in [TS playground](https://www.typescriptlang.org/play/?jsx=4&ts=5.5.0-beta#code/JYWwDg9gTgLgBAbziApiCAaOAlFBDAYxgDkIATFOAXzgDMoIQ4ByKfI5gKG9oFcA7IsAj84AWQCeAYXAAeACpwUADxgp+ZAM452JcigB8ACiQA3PABtelKgC5Ecc1ZT3FVAJSJOcOGxi8oUVkDBCdrKlkAegMAbk4qbgIRTXgxNAgUMhkwOABeZHSjSWz3OO5IyLgAdWgAa208bRUwFCJMziT+FKULAEY8uFliuX5eEAAjFCgDR0trXIQAVhposs4K6rqGpuUWtrIO5PgUCwAmAaH0zOzZUYmpmbCUBeYACxOLCGYV2O51yoACgxIJpMnBNBArDBhPxbNwKAQLHg2MhyLwLJRWLpmF4fDAJC04ABRZQQaEEKSMSD8dQwADKMDw5M0sgBAwQVBm+QQ3h8cA2+JaWDIwE0YCREmIeFQWAAJLLBSgILQlDACAA6XlUMo+PiCaEiOBIgBeEgUSlU6i0OkIMHVlPAIlp8gJKFkeH4EgMxk+eDI9iMnlyMyBjFFbqQFFoeHRMFc1AM7njADJiaTyQ7qbSGUzgAQWbhbfaqU7+DBQ2AWfJvTq6AIhIbUOhZJ1uooVGoNNpC0Ri46aWWXS13Z7vUZeT5M6W43B5BgJ3AwMDNABBNhEgCOvEsAH4A0uUKYK5p7IWyCILGae3apwPy8uFN6sDTVMfT-hz-xL7Jr32s2Xj0fRM8hmcYIEhfB+F5JNZzgVMSTJPNb2zRlmR-XQ-2nQDq1+KggA). My POC doesn't consider class components though. If I get a green-light, I can make a reasonably backward-compatible solution.

### Related discussions

See also #69029

[a comment by an automated account is not shown]

**Comment by KurtGokhan** [comment]
@eps1lon tagging you since you seem to be actively working on types. Any thoughts about this? Is there a plan to improve these types for React 19?

**Reply by eps1lon** [reply]
Not from me, no.

**Comment by 2khan** [comment]
Any updates on this?
